controllers/cruisecontroloperation_ttl_controller.go

```diff
@@ -0,0 +1,116 @@
+// Copyright © 2022 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/banzaicloud/koperator/api/v1alpha1"
+	banzaiv1alpha1 "github.com/banzaicloud/koperator/api/v1alpha1"
+)
+
+const (
+	defaultRequeueCruiseControlOperatonTTL = 30
+)
+
+// CruiseControlOperationTTLReconciler reconciles CruiseControlOperation custom resources
+type CruiseControlOperationTTLReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=kafka.banzaicloud.io,resources=cruisecontroloperations,verbs=get;list;watch;create;update;patch;delete;deletecollection
+// +kubebuilder:rbac:groups=kafka.banzaicloud.io,resources=cruisecontroloperations/status,verbs=get
+
+//nolint:gocyclo
+func (r *CruiseControlOperationTTLReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := logr.FromContextOrDiscard(ctx)
+	log.V(1).Info("reconciling CruiseControlOperation custom resources")
+
+	ccOperation := &banzaiv1alpha1.CruiseControlOperation{}
+	if err := r.Get(ctx, client.ObjectKey{Name: req.Name, Namespace: req.Namespace}, ccOperation); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			return reconciled()
+		}
+		// Error reading the object - requeue the request.
+		return requeueWithError(log, err.Error(), err)
+	}
+
+	if ccOperation.GetTTLSecondsAfterFinished() == nil || ccOperation.GetCurrentTask().Finished == nil {
+		return reconciled()
+	}
+
+	operationTTL := time.Duration(*ccOperation.GetTTLSecondsAfterFinished())
+	finishedAt := ccOperation.GetCurrentTask().Finished
+
+	if IsExpired(operationTTL, finishedAt.Time) {
+		log.Info("cleaning up finished CruiseControlOperation", "name", ccOperation.GetName(), "namespace", ccOperation.GetNamespace())
+		return r.delete(ctx, ccOperation)
+	} else {
+		return requeueAfter(int(finishedAt.Time.Add(operationTTL).Sub(time.Now())))
+	}
```

Comment:
**Recommended**: I think we **SHOULD** put this logic on function top level, because it is [more idiomatic (regular case on top level)](https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow) and thus more familiar to Go developers.